### ISSUE TITLE
RFC 0011 - Support for versioning decorators

### ISSUE THREAD

**Issue by tplooker** [hyperledger/aries-rfcs#78]
This is already marked as a pending question in [RFC-0011](https://github.com/hyperledger/aries-rfcs/tree/master/concepts/0011-decorators) but I am creating an issue to draw closer attention to it and promote discussion.

Since the conception of decorators, there have been many defined, it is now timely that we discussed strategies around how these could be versioned to allow their continual evolution.

**Draft Syntax**

A simple syntax could be to define a trailing character present after the decorators name that separates the name from a [semver base version](https://semver.org/)

```lang=json
{
 "~decorator_name/1.0" : {
    "decorator_field" : "value"
}
```

**Considerations**

At present we have a discovery protocol that supports discovering the supported message families of another party, would this protocols scope be expanded to also support the disclosure of supported decorators?

**Comment by dhh1128** [comment]
Regarding the discovery protocol: yes, I am going to update that protocol to query/return supported decorators, too.

Regarding versioning decorators: I think our need for this is real, but it's far simpler than full-blown 3-part semver. I suggest that we adopt the following conventions:

1. A decorator without a version is considered to be 1.0. (This would allow all of our current decorators to be understood to be 1.0 decorators, and 2.0 only has to be added when it becomes necessary.)
2. A decorator without a minor version is considered to be <major version>.0. (This allows `~decorator_name/2` with no need for the ".0" part).
3. Decorators never have patch versions.

**Comment by kdenhartog** [comment]
I second the 2nd and 3rd points suggested by @dhh1128

I believe that changing a decorator to require a new field in a strongly typed language would require a breaking change and would support the points of 2 and 3.

One question I would raise is do we expect that the first implementation of a decorator will always be version 1? I can see a time when it would be useful to have an experimental decorator to try a new functionality and it's been through multiple version changes by the time consensus. In this case does that mean that all implementations would have to support the versions that were created before feature parity is achieved?

**Comment by TelegramSam** [comment]
Still thinking about this. I would like to propose `~decorator.1.0` as a replacement for `~decorator/0.1`. This would eliminate . as a valid decorator name character, but I'm ok with that.

**Comment by dhh1128** [comment]
@TelegramSam The '.' as a separator is interesting because the decorator RFC already has a brief comment about using '.' for namespacing. That is, if someone writes private decorators, they should prepend their own namespace on the front: ~acme.mydecorator. Adding a dot with version (either major only, or major.minor) on the back side would be in harmony with that.

**Comment by tplooker** [comment]
Should the version be qualified in the same syntax as namespaces? I feel this would make overly complicated parsing rules if you have private version decorators. i.e

`~acme.mydecorator.1.0` <- could perhaps be harder to parse than
`~acme.mydecorator/1.0`

On the other hand we do need to keep a limited character set.

**Comment by dhh1128** [comment]
I've now raised PR #176 , which includes a [section about decorator versioning](https://github.com/hyperledger/aries-rfcs/blob/40e8d7aba2c74aadae5790d9aee8159ddfcbf84f/concepts/0011-decorators/README.md#versioning) as part of moving the decorator RFC to ACCEPTED status. If that PR is merged, I suggest we close this issue.

**Comment by tplooker** [comment]
@dhh1128 I appreciate this change and it is how I was originally thinking it best to handle versioning of this component of DIDComm. However on further thought I am wondering whether the value of independent versioning of decorators is going to be defeated by the increased scope of negotiation between parties wanting to communicate over DIDComm. 

What I propose instead is an overall messaging version for DIDComm messages that includes all decorators defined at that time of the message version release. As defined now decorators would remain optional however they would all be constrained to a collective messaging standard version. Updates to decorators and the core semantics for DIDComm would require an update to the overall messaging version used but I believe this is a healthier approach and this effective centralization of versioning would lead to a more cohesive and easier to negotiate protocol.

The middle ground I am happy to meet on, is only include the core decorators in an overall protocol version for DIDComm and allow the creation of new versioned decorators in order to not discourage innovation.

An example of the situation I am trying to prevent is the following, under the current situation (with all decorators independently version-able).

Alice and Bob want to perform credential exchange via the credential exchange protocol. 
Prior to this exchange they must negotiate what version to use of the following pieces.
- The version of credential exchange protocol
- The version of the threading decorator
- The version of the timing decorator
- The version of the attachment decorator

If the three threading decorators were apart of an overall DIDComm messaging protocol version this list of negotiation reduces to the following
- The version of credential exchange protocol
- The version of the DIDComm messaging protocol to use

**Comment by dhh1128** [comment]
@tplooker I feel like we're standing at a fork in the road, architecturally. If we go down the left fork, we end up with a DIDComm ecosystem that's statically typed and carefully pre-validated, and we build something that's pretty close to SOAP/WSDL or CORBA or DCOMM; if we go down the right fork, we end up with something that's duck typed and loosely validated, and resembles the way familiar web services (Google Maps, Github, AWS) are invoked without tight guarantees.

There are virtues to each fork, but my own judgment of the landscape is that statically typed, carefully pre-validated architectures are better for situations where a single vendor controls the whole stack. This is because a change at a lower level requires a corresponding change at a higher level, and that's painful if you have to cross team boundaries to make it happen. The fragility of that fork may be offset by the value you can get from early verification--but only if a single party controls the whole stack and really wants it to evolve in lockstep. This is not at all the case for DIDComm. I think we want in DIDComm is a world where if it walks like a duck, and quacks like a duck, it's probably going to work to treat it like a duck. Note the hedge word "probably."

If I am interested in whether a message expires, I probably don't care what version of the `~timing` decorator I'm dealing with; I just care that it has a field named `expires_time`. And I don't intend to negotiate with anybody about the version or the presence/absence of that field; I'll either see what I'm looking for, and act accordingly, or I won't see it, and I'll ignore whatever else the `~timing` decorator has that I don't care about.

That might sound a bit sloppy, but I think it would be better to characterize that attitude as pragmatic. It accepts your argument about negotiating 2 things instead of 5, and takes it one step further: most developers aren't even going to negotiate at all. They'll just send a message and see if they get an error back. Or they might query an agent capability once or twice for general guidance, but not exhaustively prove compatibility before every interaction.

I think a protocol's RFC should identify any decorators that it requires, and any fields in those decorators that are vital--but I don't think it should say something like "we require version 1.2.3 of decorator X", because that's overly inclusive. The true requirement is probably not ALL of version 1.2.3 of the timing decorator, it's just that the timing decorator have an `expires_time` field. And there's an easy way to test that without doing version negotiation: it's to just look for what you want, when the message arrives.

Stepping back from the details, the thing I'm trying to say is that if we make the versioning mechanism overly specific, we also make it overly rigid, and we end up creating more trouble than we solve. I wrote about this extensively a few years ago in a series of blog posts, if you're interested: https://codecraft.co/2008/07/29/decoupling-interfaces-as-versions-evolve-part-1/

Bottom line is that I think we already have the right sweet spot in our tradeoffs. If I haven't convinced you in this comment, then maybe we should have a chat in one of the next community calls?

**Comment by tplooker** [comment]
Posting a link to issue #126 which is associated to this discussion.

**Comment by tplooker** [comment]
@dhh1128 I appreciate the articulation and giving some examples, however I remain unconvinced at the current boundaries we have drawn. I definitely side with a protocol that is looser in its typing however at present I think we are too loose in places that is integral to DIDComm. I want to break the conversation up into a couple of questions

1. Do you believe we need an overall DIDComm message protocol version?

**Comment by tplooker** [comment]
@dhh1128 I appreciate your articulation, however I remain unconvinced about the current boundaries we have drawn in DIDComm and the future implications this has for the protocol.
I definitely side with the view that we want DIDComm to be a loosely type protocol, however it needs 
to provide enough support at the core to ensure cohesive development.

As a first point, do you think we need an overall DIDComm message protocol version?

From my point I view, I would argue we do, otherwise there is nothing that formally defines some of the core semantics of DIDComm messages such as the `@type` and `@id` fields. Without a protocol version, we would have no way of revising these semantics.

With the notion of an overall message protocol version I believe this should extend to define a set of core decorators, that are integral to DIDComm. These include the following

- Threading
- Timing (*some information such as exipry, issued at etc)
- Service
- Return Route
- Attachments

These decorators to my mind are key parts of DIDComm that provide the stability to enable broad and cohesive usage of DIDComm application layer protocols such as credential exchange. To counter your point in your previous comment, I think message expiry is an example of something that
needs a clear contract between interacting parties, otherwise this could open up several attack vectors that could undermine DIDComm. I think the fact that [JWT's](https://tools.ietf.org/html/rfc7519) define the [exp](https://tools.ietf.org/html/rfc7519#section-4.1.4) as a core field is a key demonstration of this vitality. As further evidence to this approach [UMF](https://github.com/cjus/umf) which shares a lot of similarities to DIDComm has many of the above 
conventions we presently define as optional decorators, instead defined in an overall message protocol version.

To summarize my view, I believe we should encourage the development of decorators, however a subset of those currently defined (listed above) should shift into a core message protocol version instead of being individually versioned.

**Comment by llorllale** [comment]
Perhaps a compromise can be reached by extending [RFC 0031](https://github.com/hyperledger/aries-rfcs/tree/master/features/0031-discover-features) to support multiple queries in a single roundtrip.

**Comment by dhh1128** [comment]
We just discussed this on an aries community call. Outcome = let's try to clarify the (very) small set of decorators that feel so essential that we can't have them versioned independently. Then we can write an RFC that defines and versions them all as a unit. The decorator versioning strategy proposed near the top of this thread, and currently in a PR, can then be used for other decorators that don't have such a high bar.

Others who were on the call, please chime in if I've summarized wrong.

**Comment by dhh1128** [comment]
Here is my first suggestion of what is truly core:

`~thread`
`~timing` (mainly because it has `expires_time` in it; its other features may be less important)
`~attach`?

**Comment by swcurran** [comment]
Nothing is happening on this issue and the current solution (no decorator versions) is what implementers are doing. Do we leave it like that?

Aries Stable is a new attempt at accomplishing reliable interoperability, which was discussed as a second topic in this issue.

Do we close this issue with no change?

**Comment by dhh1128** [comment]
@tplooker and I chatted about this the other day. i think we're going to close this issue with no change, but Tobias is going to suggest some other ideas that will make that outcome feel more reasonable, so I'm kind of waiting for him to socialize his ideas. Then we can close this.

**Comment by swcurran** [comment]
Can we close this issue and accept PR #176 @dhh1128 @tplooker?

As noted, I think Aries Interop Profile addresses this.

**Comment by dhh1128** [comment]
I don't think Aries Interop Profile addresses this very well, because it doesn't guarantee semver semantics for the decorators themselves--only a sort of pinning behavior--and because it assumes that the interop target people choose will be a whole profile, rather than parts of a profile. However, I'm content to close the ticket for now, if @tplooker is.

**Comment by dhh1128** [comment]
Update to above comment: I see that you actually asked 2 important and substantive questions, @swcurran--but my response only addressed half of it.

If we close this ticket without merging PR #176, then everything I said applies.

If we close the ticket and also merge PR #176, then I think it addresses the semver semantics for the decorators.

**Comment by tplooker** [comment]
Yeap I am happy to close this ticket.

**Comment by swcurran** [comment]
I'm going to change #176 to put a deadline on it for 2 weeks from now (e.g. use the normal status change mechanism) and close this issue.

I also recommend that we have a session about all the status change deadlines at the connectathon.
